MAXQ10xx Integration into wolfSSL

anhu opened the pull request:
Initial changes suggested by Anthony Hu.
This is still a work in progress. I need this to be reviewed by my peers.

This is a squashed version of https://github.com/anhu/wolfssl/pull/8 which has been closed. Review and progress will continue here.

anhu committed:
These are the initial changes from Analog Devices.

It came in the form of a patch. Note that it is no longer the same as it has
been applied on top of newer code and conflicts were resolved.

anhu committed:
Initial changes suggested by Anthony Hu.

This is still a work in progress. I need this to be reviewed by my peers.

anhu commented on wolfcrypt/src/port/maxim/maxq10xx.c:
```diff
@@ -0,0 +1,3300 @@
+        maxq10xx_make_tls_master_secret);
+    wolfSSL_CTX_SetTlsFinishedCb(ctx, maxq10xx_perform_client_finished);
+
+    wolfSSL_CTX_SetReadCertDerCb(ctx, maxq10xx_readCertDer_cb);
```
This will be replaced by an API that the application will call.

dgarske commented on src/internal.c:
```diff
@@ -2388,6 +2388,21 @@ int InitSSL_Ctx(WOLFSSL_CTX* ctx, WOLFSSL_METHOD* method, void* heap)
+        ctx->devId = MAXQ_DEVICE_ID;
+        if (method->version.major == SSLv3_MAJOR &&
+            method->version.minor == WOLFSSL_TLSV1_3) {
+            maxq10xx_SetupPkCallbacks(ctx, 1);
```
Would it be easier -> more future proof to just provide the method->version here?

anhu replied:
YES!! Stay tuned for that.

anhu replied:
Fixed by https://github.com/anhu/wolfssl/pull/9/commits/89b837a6d9312623af8ebb96218c7f6a6805c55d

dgarske commented on src/internal.c:
```diff
@@ -4571,6 +4586,12 @@ int RsaVerify(WOLFSSL* ssl, byte* in, word32 inSz, byte** out, int sigAlgo,
 
+#if defined(HAVE_PK_CALLBACKS) && defined(WOLFSSL_MAXQ108x)
+    if (ssl->options.side == WOLFSSL_CLIENT_END) {
+        maxq10xx_SetRsaPssSignature(in, inSz);
```
Why is this required? Can't the `RsaPssVerifyCb` be used? It passes in and inSz.

anhu replied:
Excellent suggestion. I will investigate.  Ah, yes, this I was going to leave until later after some consultation with the customer to understand what the hardware is actually expecting. I think the hardware is expecting the un-padded data, but need to confirm.

anhu replied:
Fixed.

dgarske commented on src/internal.c:
```diff
@@ -4984,8 +5014,10 @@ int EccVerify(WOLFSSL* ssl, const byte* in, word32 inSz, const byte* out,
-    #if !defined(WOLFSSL_RENESAS_SCEPROTECT) && \
-        !defined(WOLFSSL_RENESAS_TSIP_TLS)
+    #if defined(HAVE_PK_CALLBACKS) && defined(WOLFSSL_MAXQ108x)
+    if ((ret == NOT_COMPILED_IN) || ssl->options.side != WOLFSSL_CLIENT_END)
```
What is this? Fallback to software if not client  or it returns NOT_COMPILED_IN? Why only client?

anhu replied:
> Fallback to software if it returns NOT_COMPILED_IN?

Yes. I think this is a rather nice.  When wolfssl calls a callback, wolfssl doesn't know whether or not the callback actually supports the operation.  For example, an operation might support certain curves (P256) but not other ones (P521) so if it  returns `NOT_COMPILED_IN` then that signals a fallback to wolfssl's implementation.  Does this seem reasonable?

> Why only client?

Ah this is a left over artifact from the way the customer gave us the patch.  The MAXQ hardware does not support being the server; only being the client. So they only want the maxq10xx APIs called if this is the client side. I've had discussions with the customer and we came up with a better approach where when support for MAXQ is turned on at configure time, then support for being the server side of a TLS connection will be disabled.  When I make that change I will also remove all these checks for `WOLFSSL_CLIENT_END`.  That will likely come after I've done all the possible callback conversions. 

For server side functionality, a user would simply use another instance of wolfssl.

anhu replied:
"Only client" fixed. Keeping fallback.

dgarske commented on src/internal.c:
```diff
@@ -5237,7 +5269,11 @@ int Ed25519Sign(WOLFSSL* ssl, const byte* in, word32 inSz, byte* out,
             keySz, ctx);
     }
+#if defined(HAVE_PK_CALLBACKS) && defined(WOLFSSL_MAXQ108x)
+    if ((ret == NOT_COMPILED_IN) || ssl->options.side != WOLFSSL_CLIENT_END)
```
Same why only client?

anhu replied:
See above.  Will be fixed to not check for client side.

anhu replied:
Fixed. Keeping fallback.

dgarske commented on src/internal.c:
```diff
@@ -13656,7 +13714,8 @@ int ProcessPeerCerts(WOLFSSL* ssl, byte* input, word32* inOutIdx,
 
-                #ifndef OPENSSL_COMPATIBLE_DEFAULTS
+                #if !defined(OPENSSL_COMPATIBLE_DEFAULTS) && \
+                    !defined(DISABLE_PEER_CERT_VERS_CHECK)
```
Please clearly document `DISABLE_PEER_CERT_VERS_CHECK`

anhu replied:
Removed.  It is guarding code that is only executed on the server side.

dgarske commented on src/internal.c:
```diff
@@ -26961,6 +27054,16 @@ static int HashSkeData(WOLFSSL* ssl, enum wc_HashType hashType,
     #endif
 
+    #if defined(HAVE_PK_CALLBACKS)
+        if (ssl->options.side == WOLFSSL_CLIENT_END && ssl->ctx &&
```
Again why only client?

anhu replied:
Fixed.

dgarske commented on src/internal.c:
```diff
@@ -26983,6 +27086,12 @@ static int HashSkeData(WOLFSSL* ssl, enum wc_HashType hashType,
 
+    #ifdef WOLFSSL_MAXQ10XX_TLS
+        if (ssl->maxq_ctx.device_cert) {
+            ssl->options.sendVerify = SEND_CERT;
```
What is this? Is it forcing mutual auth? I don't like it. We have API's to force mutual auth.

anhu replied:
Removed.

dgarske commented on src/internal.c:
```diff
@@ -28823,20 +28948,48 @@ int SendClientKeyExchange(WOLFSSL* ssl)
+                            ERROR_OUT(PSK_KEY_ERROR, exit_scke);
+                        }
+
+                        // save client_identity
```
C style `/* */` only please.

anhu replied:
All instances of // added by this PR have been changed to `/*  */`

dgarske commented on src/internal.c:
```diff
@@ -29285,13 +29438,21 @@ int SendClientKeyExchange(WOLFSSL* ssl)
+                    #ifdef WOLFSSL_MAXQ10XX_TLS
+                    if ((ssl->hsKey) &&
+                        (((ecc_key*)ssl->hsKey)->maxq_ctx.hw_storage == 1)) {
+                        WOLFSSL_MSG("Skipping EccSharedSecret()");
```
Skipping shared secret calc because it is done in. HW? Can we do this in a more generic way like a certain new PK callback is setup?

anhu replied:
I came up with the idea of setting `EccSharedSecretCb` to a callback that does nothing.  But there are a few instances of : 
```
                    if (ssl->ctx->EccSharedSecretCb != NULL) {
                        break;
                    }
```

Which made it really confusing and hard to debug. 

I should have read your comment more carefully.  You are suggesting a new callback.  I can add a new one called `EccSkipSharedSecret` . It would indicate whether or not to call `EccSharedSecret()`

anhu replied:
New callback created.

dgarske commented on src/tls.c:
```diff
@@ -566,11 +560,38 @@ int MakeTlsMasterSecret(WOLFSSL* ssl)
-              ssl->heap, ssl->devId);
+        {
+#ifdef HAVE_PK_CALLBACKS
+            if ((ssl->options.side == WOLFSSL_CLIENT_END) &&
```
Again client only?

anhu replied:
fixed.

dgarske commented on src/tls13.c:
```diff
@@ -165,6 +165,89 @@ static const byte tls13ProtocolLabel[TLS13_PROTOCOL_LABEL_SZ + 1] = "tls13 ";
+ * extend beyond the initialized span.
+ */
+PRAGMA_GCC_DIAG_PUSH;
+PRAGMA_GCC("GCC diagnostic ignored \"-Wmaybe-uninitialized\"");
```
Please investigate this. I don't like suppressing the warning unless we are 100% clear why and it is clearly documented.

anhu replied:
This is brought over from above.  I think it was added by Douzzer.

dgarske commented on wolfcrypt/src/hmac.c:
```diff
@@ -341,6 +341,14 @@ int wc_HmacSetKey(Hmac* hmac, int type, const byte* key, word32 length)
 
+#if defined(HAVE_PK_CALLBACKS) && defined(WOLFSSL_MAXQ108x)
+    /* Nothing left to do here. */
+    /* TODO: this makes the server fail. Is that okay?
```
Investigate this...

anhu replied:
Removed.

dgarske commented on wolfcrypt/src/random.c:
```diff
@@ -838,6 +838,11 @@ static int _InitRng(WC_RNG* rng, byte* nonce, word32 nonceSz,
     rng->status = DRBG_NOT_INIT;
 #endif
 
+#if defined (WOLF_CRYPTO_CB) && defined(WOLFSSL_MAXQ108x)
```
Loose space `defined(WOLF_CRYPTO_CB)`. Hard coding when `WOLFSSL_MAXQ108x` is defined is a bit of a hack. Technically you can set all this via TLS using `wolfSSL_SetDevId` at run-time. I saw another place this was hard coded. Also is it only the WOLFSSL_MAXQ108x the part that supports RNG?

anhu replied:
I think they do this in case a user wanted to use wolfcrypt without wolfssl. Maybe I can figure out something better.  I should test if this is even required anymore. I've asked if 1065 also supports RNG.

anhu replied:
Removed.  Its already set to the correct devid via changes I made.

dgarske commented on wolfcrypt/src/wc_port.c:
```diff
@@ -2178,6 +2189,221 @@ int wolfSSL_CryptHwMutexUnLock(void)
     }
 
+#elif defined(MAXQ10XX_MUTEX)
+    static pthread_mutex_t *wcCryptHwSharedMutexPtr;
```
Section is not portable. Fine with me as long as this `MAXQ10XX_MUTEX` can be disabled. Also noticed not using our stdlib XMALLOC/XMEMSET, etc.

anhu replied:
`MAXQ10XX_MUTEX` is only enabled if configured with `--with-maxq10xx`

dgarske commented on wolfssl/internal.h:
```diff
@@ -1654,6 +1654,15 @@ enum Misc {
 
+#ifndef WOLFSSL_NO_TLS12
+#ifdef WOLFSSL_SHA384
+    #define HSHASH_SZ WC_SHA384_DIGEST_SIZE
```
We likely already have a macro or enum for this. If not please document what this value is and put with other TLS related settings in this file.

anhu replied:
Its moved from `src/tls.c` to here, but I now noticed it is no longer required. Will move it back.

anhu replied:
Done.

dgarske commented on wolfssl/ssl.h:
```diff
@@ -3505,7 +3511,7 @@ WOLFSSL_API void* wolfSSL_GetEncryptKeysCtx(WOLFSSL* ssl);
 typedef int (*CallbackTlsFinished)(WOLFSSL* ssl,
                             const byte *side,
-                            const byte *handshake_hash,
+                            const byte *handshake_hash, word32 hashSz,
```
I think we can take this API breakage... `CallbackTlsFinished` But need to confirm.

anhu replied:
Thanks!! Yes. We talked about this over internal chat. Is there someone else I need to check with? The change does make sense as the hash size is not a constant.

dgarske commented on wolfssl/wolfcrypt/port/maxim/maxq10xx.h:
```diff
@@ -0,0 +1,142 @@
+/* maxq10xx.h
```
I don't see where this new file is handled in an include.am.

anhu replied:
Ah!! Thank you! This might solve a problem I've been seeing!!

anhu replied:
Fixed by https://github.com/anhu/wolfssl/pull/9/commits/e74d258c87e7db9d5c1827c71dad8530751137de .

anhu committed:
Actually install the maxq10xx.h header

anhu committed:
Pass in ProtocolVersion to maxq10xx_SetupPkCallbacks

anhu committed:
Get rid of MakeTlsMasterSecretCb. Redundancy.

anhu commented on src/internal.c:
```diff
@@ -28063,17 +28166,33 @@ static int DoServerKeyExchange(WOLFSSL* ssl, const byte* input,
-                            );
+                        #ifdef HAVE_PK_CALLBACKS
+                            ret = NOT_COMPILED_IN;
+                            if (ssl->options.side == WOLFSSL_CLIENT_END &&
+                                ssl->ctx && ssl->ctx->ProcessServerKexCb) {
+                                ret = ssl->ctx->ProcessServerKexCb(ssl,
+                                    args->sigAlgo,
+                                    args->verifySig, args->verifySigSz,
+                                    ssl->buffers.sig.buffer, SEED_LEN,
+                                    &ssl->buffers.sig.buffer[SEED_LEN],
+                                    (ssl->buffers.sig.length - SEED_LEN));
+                            }
+
+                            if (ret == NOT_COMPILED_IN)
+                        #endif /* HAVE_PK_CALLBACKS */
+                            {
+                                ret = EccVerify(ssl,
```
TODO: This requires a bit more testing and investigation. Its saying "Prepare for server key exchange (keygen?) and if successful, skip over ECDSA verification. Very strange.

anhu replied:
Created `EccSkipSharedSecretCb` to accommodate this.  The callback should be renamed to indicate both signature  verification and shared secret stuff is happening here.

anhu replied:
Fixed in 57ef02d6f2ea94710ac19d854756a1d95e75464d

haydenroche5 commented on configure.ac:
```diff
@@ -1869,6 +1869,39 @@ then
+            ENABLED_MAXQ10XX="yes"
+        fi
+
+        if test "$ENABLED_MAXQ10XX" = "no"; then
```
I don't think the logic here handles invalid part numbers. I suggest a switch with cases and have a default case that rejects any uncovered cases.

anhu replied:
Fixed with ea3737b7d548fa3db5675fb5b243ddea69d5b0a4

haydenroche5 commented on configure.ac:
```diff
@@ -1869,6 +1869,39 @@ then
+        if test "$maxqpartnumber" = "MAXQ108x"; then
+            LIB_ADD="$LIB_ADD -lrt"
+            LIB_STATIC_ADD="$LIB_STATIC_ADD lib/libmaxq108x_api.a"
+            AM_CFLAGS="$AM_CFLAGS -DWOLFSSL_MAXQ108x"
```
Prefer all caps for macros. So WOLFSSL_MAXQ108X.

anhu replied:
Fixed with aafaa6d3f35abc14706b831339b4626428c2d876

haydenroche5 commented on configure.ac:
```diff
@@ -1869,6 +1869,39 @@ then
+        maxqpartnumber=$withval
+
+        if test "$maxqpartnumber" = "MAXQ1065"; then
+            LIB_ADD="$LIB_ADD -lrt"
```
Should be able to use AC_CHECK_LIB or similar to check for librt.

anhu replied:
Fixed with ea3737b7d548fa3db5675fb5b243ddea69d5b0a4

anhu committed:
Get rid of calls to maxq10xx_SetRsaPssSignature and maxq10xx_RsaPssVerify.

anhu committed:
Cleanup of maxq10xx.h

anhu committed:
Various Changes.

- rename maxq10xx_readCertDer_cb() to wolfSSL_maxq10xx_load_certificate()
- call wolfSSL_maxq10xx_load_certificate() in InitSSL(); instead of callback
- maxq10xx.h cleanup: whitespace, WOLFSSL_LOCAL, extern "C"

anhu committed:
Temporarily add new example client code here.

Build:

gcc -o examples/maxq10xx-wolfssl-client examples/maxq10xx-wolfssl-client.c -Wall -I./ -Os -L./src/.libs/ -lm -lwolfssl

Execute:

LD_LIBRARY_PATH=src/.libs/ ./examples/maxq10xx-wolfssl-client

anhu committed:
Get rid of ReadCertDerCb. No longer needed.

anhu committed:
include config.h

anhu committed:
Enforce client side-only at build-time...

...and stop checking for side being client at run time.
Enforcement is via #error macro directive.

anhu committed:
Correct a small macro error

haydenroche5 commented on examples/maxq10xx-wolfssl-client.c:
```diff
@@ -0,0 +1,335 @@
+/* maxq10xx-wolfssl-client.c
```
For something specific to a particular piece of hardware like this, I wonder if the examples repo is a better location.

anhu replied:
ah, Yes. I think so as well. This was put here temporarily so the customer wouldn't have to look in the examples. repo.  This will be removed before final merge.

dgarske replied:
I agree with Hayden on this. Please make sure this example is moved to wolfssl-examples and removed from here before marking this PR ready. Let me know when you'd like me to review again.

anhu replied:
Yes, I've already done it locally.

anhu replied:
Done with 7d04d275f

haydenroche5 commented on examples/maxq10xx-wolfssl-client.c:
```diff
@@ -0,0 +1,335 @@
+/* Configurations */
+/* -------------- */
+
+/* Please define this if you want wolfSSL's debug output */
```
Just personal preference, but I don't think we need to say please for all our comments. :)

anhu replied:
It signals that the reader has something they need to do.

haydenroche5 commented on examples/maxq10xx-wolfssl-client.c:
```diff
@@ -0,0 +1,335 @@
+        key[2] = 0x3c;
+        key[3] = 0x4d;
+
+        return 4;   /* length of key in octets or 0 for error */
```
Move this comment above the function signature. Then you don't need to repeat yourself at line 119.

anhu replied:
this code has been removed .

haydenroche5 commented on examples/maxq10xx-wolfssl-client.c:
```diff
@@ -0,0 +1,335 @@
+        int i;
+        int b = 0x01;
+
+        for (i = 0; i < 32; i++, b += 0x22) {
```
Instead of using 32 and 4, would prefer named constants that tell me what these are (i.e. avoid magic numbers).

anhu replied:
This is taken from test.h. Would prefer to keep it the same.

haydenroche5 commented on examples/maxq10xx-wolfssl-client.c:
```diff
@@ -0,0 +1,335 @@
+        int b = 0x01;
+
+        for (i = 0; i < 32; i++, b += 0x22) {
+            if (b >= 0x100)
```
I don't really understand the logic of this loop. Explanatory comment would help.

anhu replied:
This is taken from test.h. Would prefer to keep it the same.

haydenroche5 commented on examples/maxq10xx-wolfssl-client.c:
```diff
@@ -0,0 +1,335 @@
+    if (inet_pton(AF_INET, DEFAULT_SERVER, &servAddr.sin_addr) != 1) {
+        fprintf(stderr, "ERROR: invalid address\n");
+        ret = -1;
+        goto exit;
```
I know this is just example code, but would still prefer we stick to our coding standard and not use gotos.

anhu replied:
This is taken from test.h. Would prefer to keep it the same.

haydenroche5 commented on src/internal.c:
```diff
@@ -2388,6 +2388,12 @@ int InitSSL_Ctx(WOLFSSL_CTX* ctx, WOLFSSL_METHOD* method, void* heap)
 #endif /* HAVE_WOLF_EVENT */
 
+#if defined(HAVE_PK_CALLBACKS) && defined(WOLFSSL_MAXQ10XX_TLS)
+    /* Let the maxq10xx infrastructure know whether we are doing TLS 1.3. */
```
I would change this comment to "Let the maxq10xx infrastructure know what TLS version we're using."

anhu replied:
fixed in 5c530b1a4

haydenroche5 commented on src/internal.c:
```diff
@@ -5818,6 +5830,15 @@ int DhGenKeyPair(WOLFSSL* ssl, DhKey* dhKey,
+    if (ssl && ssl->ctx && ssl->ctx->DhGenerateKeyPairCb) {
+        ret = ssl->ctx->DhGenerateKeyPairCb(dhKey, ssl->rng, priv, privSz,
+                                            pub, pubSz);
+        if (ret != NOT_COMPILED_IN)
```
What about when ret == NOT_COMPILED_IN? Do we want to fall back to wc_DhGenerateKeyPair? Because that's what'll happen as written.

anhu replied:
Yes.  I think so.  This gives hardware implementations a chance to validate that they can support the operation. For example, if the hardware does not support `ffdhe8192` then this gracefully allow fallback to wolfssl's implementation.

haydenroche5 commented on src/internal.c:
```diff
@@ -7047,6 +7068,11 @@ int InitSSL(WOLFSSL* ssl, WOLFSSL_CTX* ctx, int writeDup)
 #endif
+
+#if defined(WOLFSSL_MAXQ10XX_TLS)
+    wolfSSL_maxq10xx_load_certificate(ssl);
```
Check return value.

anhu replied:
Fixed in e529b95464797fba459ff2f0dbbad1c63b0d7444

haydenroche5 commented on src/internal.c:
```diff
@@ -12913,6 +12939,30 @@ static int ProcessPeerCertParse(WOLFSSL* ssl, ProcPeerCertArgs* args,
 #endif
 
+#ifdef WOLFSSL_MAXQ10XX_TLS
+    /* MAXQ10xx accepts ASN_NO_SIGNER_E because the CA cert is registered with
```
This seems like a case that should be coded into the cert manager logic, rather than carving out an exception here for it, but I'm not familiar with all the details.

anhu replied:
Not sure what that would look like.  I'm going to leave it as is for now.

anhu replied:
Huh, this logic isn't actually required as the maxq callback will always override the return value later.  Removing this logic.  Thanks for making me take a closer look.

anhu replied:
Removed in 2f06dede275d8c7d1924896d1ee358da840f0104

haydenroche5 commented on src/internal.c:
```diff
@@ -12913,6 +12939,30 @@ static int ProcessPeerCertParse(WOLFSSL* ssl, ProcPeerCertArgs* args,
+    }
+#endif /* WOLFSSL_MAXQ10XX_TLS */
+
+#ifdef HAVE_PK_CALLBACKS
```
Please add a comment explaining what this block is doing.

anhu replied:
Fixed with 3b5a8330c

haydenroche5 commented on src/internal.c:
```diff
@@ -13656,7 +13706,8 @@ int ProcessPeerCerts(WOLFSSL* ssl, byte* input, word32* inOutIdx,
 
-                #ifndef OPENSSL_COMPATIBLE_DEFAULTS
+                #if !defined(OPENSSL_COMPATIBLE_DEFAULTS) && \
+                    !defined(DISABLE_PEER_CERT_VERS_CHECK)
```
DISABLE_PEER_CERT_VERS_CHECK needs a WOLFSSL prefix. Please don't add new macros without namespacing.

anhu replied:
This code has been removed.

haydenroche5 commented on src/internal.c:
```diff
@@ -17249,13 +17300,30 @@ static WC_INLINE int EncryptDo(WOLFSSL* ssl, byte* out, const byte* input,
+                         ssl->encrypt.additional, AEAD_AUTH_DATA_SZ);
+            }
+
+            if (ret == NOT_COMPILED_IN)
```
Here's that fallback behavior again. If NOT_COMPILED_IN, we use our default implementation. Is this how we want it to work?

anhu replied:
Yes, I think so.  This allows for validation of support.

haydenroche5 commented on src/internal.c:
```diff
@@ -28823,20 +28928,46 @@ int SendClientKeyExchange(WOLFSSL* ssl)
+                                 ssl->arrays->client_identity,
+                                 MAX_PSK_ID_LEN, dummy_psk,
+                                 MAX_PSK_KEY_LEN);
+                    if (ret_cb == 0 || ret_cb > MAX_PSK_KEY_LEN) {
```
Does ret_cb have the length? If so, I think you should be putting the \0 at line 28945 just after the length (i.e. after the first ret_cb bytes), not at the end of the buffer.

anhu replied:
The code is modelled after wolfSSL code just below it.  Will keep it the same. However, I might reformat to make it look more similar.

anhu replied:
This code has been removed and replaced with generic return  code logic.  This is in 7d04d275f

haydenroche5 commented on src/internal.c:
```diff
@@ -29285,13 +29416,21 @@ int SendClientKeyExchange(WOLFSSL* ssl)
+                    #ifdef WOLFSSL_MAXQ10XX_TLS
+                    if ((ssl->hsKey) &&
+                        (((ecc_key*)ssl->hsKey)->maxq_ctx.hw_storage == 1)) {
+                        WOLFSSL_MSG("Skipping EccSharedSecret()");
```
I would modify this log to indicate that it's being skipped because WOLFSSL_MAXQ10XX_TLS is defined. Just more descriptive.

anhu replied:
This code has been removed.

haydenroche5 reviewed (changes requested):
First batch of comments.

anhu commented on wolfcrypt/src/port/maxim/maxq10xx.c:
```diff
@@ -0,0 +1,3352 @@
+            maxq10xx_perform_tls12_record_processing);
+    }
+
+    wolfSSL_CTX_SetProcessServerCertCb(ctx,
```
This callback needs to be renamed.  `Server` ----> `Peer` .    All associated pointers and types and functions need to be renamed as well.

anhu replied:
Done in 3b5a8330c

anhu committed:
Don't skip verification and simply fallback to wolfssl.

anhu committed:
Revert an unneccessary change.

anhu committed:
Revert experimental change.

anhu committed:
No longer need to hardcode here; id already properly set.

anhu committed:
New callback: EccSkipSharedSecretCb

anhu committed:
Remove DISABLE_PEER_CERT_VERS_CHECK as this only affects server side. Dead code.

haydenroche5 commented on wolfcrypt/src/aes.c:
```diff
@@ -2838,6 +2838,12 @@ static WARN_UNUSED_RESULT int wc_AesDecrypt(
     #endif
 
+    #ifdef WOLFSSL_MAXQ10XX_CRYPTO
+        if (wc_MAXQ10XX_AesSetKey(aes, userKey, keylen) != 0) {
```
I always prefer to return the error code of the function that failed, rather than converting it to a different one (WC_HW_E in this case). Just makes debugging easier if you know the original return code of the failing function.

anhu replied:
Convention in this function is to return WC_HW_E . Going to stick with it.

haydenroche5 commented on wolfcrypt/src/hmac.c:
```diff
@@ -341,6 +341,11 @@ int wc_HmacSetKey(Hmac* hmac, int type, const byte* key, word32 length)
     hmac->keyLen = length;
 #endif
 
+#if defined(HAVE_PK_CALLBACKS) && defined(WOLFSSL_MAXQ108x)
```
Is there a case where WOLFSSL_MAXQ108x would be defined without HAVE_PK_CALLBACKS? If not, you could simplify this gate.

anhu replied:
I've asked the customer. Awaiting a response.

anhu replied:
Approved by customer via email. Will work on this next.

anhu replied:
Done in c70ceb79d

haydenroche5 commented on wolfcrypt/src/wc_port.c:
```diff
@@ -257,6 +261,14 @@ int wolfCrypt_Init(void)
+    #ifdef MAXQ10XX_MODULE_INIT
+        ret = maxq10xx_port_init();
+        if (ret != 0) {
+            WOLFSSL_MSG("MAXQ10xx port init failed");
```
Prefer WOLFSSL_ERROR_MSG. Trying to encourage its use for error messages.

anhu replied:
This would look incosistent if we sometimes use WOLFSSL_MSG and other times use WOLFSSL_ERROR_MSG.  

Ah, but I can do it in `maxq10xx.c` and be consistent through the whole file. I will do that.

anhu replied:
Done in f23cd2a3e .

haydenroche5 commented on wolfcrypt/src/wc_port.c:
```diff
@@ -2178,6 +2190,221 @@ int wolfSSL_CryptHwMutexUnLock(void)
+            if (pthread_setspecific(key_own_hw_mutex, key_ptr)) {
+                return THREAD_STORE_SET_E;
+            }
+        } else {
```
Alignment and else goes on a newline. There are other instances of this in this PR, too.

anhu replied:
Fixed in eb17d3b9f.

anhu committed:
Stop forcing mutual in library code authentication.

anhu commented:
At this point all of @dgarske 's comments have been addressed.

anhu committed:
WOLFSSL_MAXQ108x ---> WOLFSSL_MAXQ108X

anhu committed:
Fixups in configure.ac

anhu committed:
Check a return code.

anhu committed:
ProcessServerCertCb ---> ProcessPeerCertCb. Also add a comment explaining it.

anhu committed:
Reformat to look more like wolfssl code

anhu committed:
Remove unneccessary check.

anhu commented on src/internal.c:
```diff
@@ -28823,20 +28920,46 @@ int SendClientKeyExchange(WOLFSSL* ssl)
                 case psk_kea:
                 {
                     byte* pms = ssl->arrays->preMasterSecret;
+                #ifdef WOLFSSL_MAXQ10XX_TLS
+                    /* For MAXQ, the hardware stores the PSK so we don't need it
+                     * from the callback. The hardware also generates the pre
+                     * master secret so we don't need it. */
+                    byte dummy_psk[MAX_PSK_KEY_LEN];
+                    word32 ret_cb;
+                    /* get client_identity */
+                    ret_cb = ssl->options.client_psk_cb(ssl,
+                        ssl->arrays->server_hint, ssl->arrays->client_identity,
+                        MAX_PSK_ID_LEN, dummy_psk, MAX_PSK_KEY_LEN);
+                    if (ret_cb == 0 || ret_cb > MAX_PSK_KEY_LEN) {
+                        ERROR_OUT(PSK_KEY_ERROR, exit_scke);
+                    }
+
+                    /* Ensure the buffer is null-terminated. */
+                    ssl->arrays->client_identity[MAX_PSK_ID_LEN] = '\0';
+                    args->encSz = (word32)XSTRLEN(ssl->arrays->client_identity);
+                    if (args->encSz > MAX_PSK_ID_LEN) {
+                         ERROR_OUT(CLIENT_ID_ERROR, exit_scke);
+                    }
+                    XMEMCPY(args->encSecret, ssl->arrays->client_identity,
+                            args->encSz);
+                    break;
+                #endif /* WOLFSSL_MAXQ10XX_TLS */
+
```
@dgarske , would like your opinion.  Should this be a callback?

dgarske replied:
The PSK client is already a callback. This looks identical to the section of code below?

anhu replied:
fixed in 7d04d275f920f0edee5f665c4cdeb9d37ad5a9bb

anhu committed:
Whitespace and formatting.

anhu committed:
use WOLFSSL_ERROR_MSG in maxq10xx

anhu committed:
Revert a change.

anhu committed:
Move maxq10xx-wolfssl-client.c into wolfssl-examples repo.

anhu committed:
Replace MAXQ code block with some generic return code logic.

anhu commented:
With the new PSK callback return code logic, this works with the new MAXQ10xx sample in this PR: https://github.com/wolfSSL/wolfssl-examples/pull/345

anhu committed:
Improve a comment.

anhu committed:
Turn on pkcallbacks automatically.

anhu committed:
whitespace

anhu committed:
ProcessServerKexCb --> ProcessServerSigKexCb

And bring back fallback to software verification.

anhu commented:
Comments have been addressed.  Closing this PR, will  squash, rebase and and open a new PR in the wolfssl/wolfssl repo for final review.
